src/Bedrock.Framework/Protocols/Protocol.cs

```diff
@@ -87,89 +89,89 @@ public ProtocolReader(ConnectionContext connection, IProtocolReader<TReadMessage
 
         public ConnectionContext Connection { get; }
 
-        public async ValueTask<TReadMessage> ReadAsync(CancellationToken cancellationToken = default)
+        public async ValueTask<ReadResult<TReadMessage>> ReadAsync(CancellationToken cancellationToken = default)
         {
             var input = Connection.Transport.Input;
             var reader = _reader;
 
             TReadMessage protocolMessage = default;
+            var isCanceled = false;
+            var isCompleted = false;
 
             while (true)
             {
                 var result = await input.ReadAsync(cancellationToken);
+                isCanceled = result.IsCanceled;
+                isCompleted = result.IsCompleted;
                 var buffer = result.Buffer;
-                var consumed = buffer.Start;
-                var examined = buffer.End;
+                _consumed = buffer.Start;
+                _examined = buffer.End;
+
+
+                if (result.IsCanceled)
+                {
+                    break;
```

Comment:
isCanceled = true here

Reply:
It's assigned before this so all results should have the right IsCanceled and IsCompleted